File: app/src/main/java/com/flysolo/dmmsugradelevelapp/views/auth/LoginActivity.java
package com.flysolo.dmmsugradelevelapp.views.auth;

import androidx.appcompat.app.AppCompatActivity;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.flysolo.dmmsugradelevelapp.R;
import com.flysolo.dmmsugradelevelapp.databinding.ActivityLoginBinding;
import com.flysolo.dmmsugradelevelapp.model.Accounts;
import com.flysolo.dmmsugradelevelapp.model.UserType;
import com.flysolo.dmmsugradelevelapp.services.auth.AuthService;
import com.flysolo.dmmsugradelevelapp.services.auth.AuthServiceImpl;
import com.flysolo.dmmsugradelevelapp.utils.LoadingDialog;
import com.flysolo.dmmsugradelevelapp.utils.UiState;
import com.flysolo.dmmsugradelevelapp.utils.Validation;
import com.flysolo.dmmsugradelevelapp.views.dialogs.ChooseClassDialog;
import com.flysolo.dmmsugradelevelapp.views.dialogs.ResetPasswordDialog;
import com.flysolo.dmmsugradelevelapp.views.student.StudentMainActivity;
import com.flysolo.dmmsugradelevelapp.views.teacher.TeacherMainActivity;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;

public class LoginActivity extends AppCompatActivity {
    private ActivityLoginBinding binding;
    private final Validation validation = new Validation();
    private final LoadingDialog loadingDialog = new LoadingDialog(this);
    private final AuthServiceImpl authService =new AuthServiceImpl(FirebaseAuth.getInstance(), FirebaseFirestore.getInstance(), FirebaseStorage.getInstance());
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.buttonLogin.setOnClickListener(view -> {
          String email = binding.inputEmail.getText().toString();
          String password = binding.inputPassword.getText().toString();
          if (!validation.isValidEmail(email) || email.isEmpty()) {
              binding.inputEmail.setError("Invalid email");
          } else if (password.isEmpty()) {
              binding.inputPassword.setError("Invalid password");
          } else {
              login(email,password);
          }
        });
        binding.buttonForgotPassword.setOnClickListener(view -> {
            ResetPasswordDialog resetPasswordDialog = new ResetPasswordDialog();
            if (!resetPasswordDialog.isAdded()) {
                resetPasswordDialog.show(getSupportFragmentManager(),"Reset Password");
            }
        });
        binding.buttonRegister.setOnClickListener(view -> startActivity(new Intent(LoginActivity.this,RegisterActivity.class)));
    }

    private void login(String email ,String password) {

        authService.login(email, password, new UiState<FirebaseUser>() {
            @Override
            public void Loading() {
                loadingDialog.showLoadingDialog("Logging in...");
            }

            @Override
            public void Successful(FirebaseUser data) {
                loadingDialog.stopLoading();
                Toast.makeText(LoginActivity.this, "Successfully Logged in!", Toast.LENGTH_SHORT).show();
                getUserInfo(data.getUid());
            }

            @Override
            public void Failed(String message) {
                loadingDialog.stopLoading();
                Toast.makeText(LoginActivity.this, message, Toast.LENGTH_SHORT).show();
            }
        });
    }
    private void getUserInfo(String uid) {
        authService.getAccount(uid, new UiState<Accounts>() {
            @Override
            public void Loading() {
               loadingDialog.showLoadingDialog("Identifying user...");
            }

            @Override
            public void Successful(Accounts data) {
                loadingDialog.stopLoading();
                if (data.getType() == UserType.STUDENT) {
//                    ChooseClassDialog classDialog = ChooseClassDialog.newInstance(data.getId());
//                    if (!classDialog.isAdded()){
//                        classDialog.show(getSupportFragmentManager(),"Choose class");
//
                    startActivity(new Intent(LoginActivity.this, StudentMainActivity.class));
                } else if(data.getType() == UserType.TEACHER) {
                    startActivity(new Intent(LoginActivity.this, TeacherMainActivity.class));
                } else {
                    Toast.makeText(LoginActivity.this, "unidentified user!", Toast.LENGTH_SHORT).show();
                    FirebaseAuth.getInstance().signOut();
                }
            }

            @Override
            public void Failed(String message) {
                loadingDialog.stopLoading();
                new MaterialAlertDialogBuilder(binding.getRoot().getContext())
                        .setMessage(message)
                        .show();
                Toast.makeText(LoginActivity.this, message, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        FirebaseUser user =FirebaseAuth.getInstance().getCurrentUser();
        if (user != null) {
            getUserInfo(user.getUid());
        }
    }
}

File: app/src/main/java/com/flysolo/myapplication/views/auth/LoginActivity.kt
package com.flysolo.myapplication.views.auth

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import com.flysolo.myapplication.utils.UiState
import com.danica.ikidsv2.utils.Validation
import com.flysolo.myapplication.databinding.ActivityLoginBinding
import com.flysolo.myapplication.models.Accounts
import com.flysolo.myapplication.models.UserType
import com.flysolo.myapplication.services.auth.AuthServiceImpl
import com.flysolo.myapplication.utils.LoadingDialog
import com.flysolo.myapplication.views.student.StudentmainActivity
import com.flysolo.myapplication.views.teacher.TeacherMainActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.FirebaseStorage

class LoginActivity : AppCompatActivity() {
    private lateinit var binding : ActivityLoginBinding
    private lateinit var authService : AuthServiceImpl
    private val  loadingDialog = LoadingDialog(this)
    private val validation = Validation()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)
        authService = AuthServiceImpl(FirebaseAuth.getInstance(), FirebaseFirestore.getInstance(), FirebaseStorage.getInstance())
        binding.buttonLogin.setOnClickListener {
            val email = binding.inputEmail.text.toString()
            val password = binding.inputPassword.text.toString()
            if (!validation.validateEmail(binding.layoutEmail) || !validation.validatePassword(binding.layoutPassword)) {
                return@setOnClickListener
            } else {
                login(email = email,password = password)
            }
        }
        binding.buttonRegister.setOnClickListener {
            startActivity(Intent(this,RegisterActivity::class.java))
        }
    }
    private fun login(email : String ,password : String) {
        authService.login(email,password) {
            when(it) {
                is UiState.Failed -> {
                    loadingDialog.stopLoading()
                    Toast.makeText(this,it.message,Toast.LENGTH_SHORT).show()
                }
                UiState.Loading -> {
                    loadingDialog.showLoadingDialog("Logging in....")
                }
                is UiState.Successful -> {
                    loadingDialog.stopLoading()
                    Toast.makeText(this,"Successfully Logged in!",Toast.LENGTH_SHORT).show()
                    getUserInfo(it.data.uid)
                }
            }
        }
    }
    private fun getUserInfo(uid : String) {
        authService.getUserInfo(uid) {
            when(it) {
                is UiState.Failed -> {
                    loadingDialog.stopLoading()
                    Toast.makeText(this,it.message,Toast.LENGTH_SHORT).show()
                }
                UiState.Loading -> {
                    loadingDialog.showLoadingDialog("Identifying user....")
                }
                is UiState.Successful -> {
                    loadingDialog.stopLoading()
                    updateUI(it.data)
                }
            }
        }
    }
    private fun updateUI(accounts: Accounts) {
        when (accounts.type) {
            UserType.TEACHER -> {
                startActivity(Intent(this,TeacherMainActivity::class.java).putExtra("user",accounts))
            }
            UserType.STUDENTS -> {
                startActivity(Intent(this,StudentmainActivity::class.java).putExtra("user",accounts))
            }
            else -> {
                Toast.makeText(this,"User not found!",Toast.LENGTH_SHORT).show()
                FirebaseAuth.getInstance().signOut()
            }
        }
    }

    override fun onStart() {
        super.onStart()
        val currentUser = FirebaseAuth.getInstance().currentUser
        if (currentUser!= null) {
            getUserInfo(currentUser.uid)
        }
    }

}